experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,654 @@
+    
+    NullConeIdeal::MPolyIdeal
+    
+    #This should be the only one that is used
+    function InvariantRing(R::RepresentationReductiveGroup) #here G already contains information n and rep_mat
+        z = new()
+        n = ncols(R.rep_mat)
+        z.representation = R
+        z.group = R.group
+        G = z.group
+        z.field = G.field
+        z.poly_ring, __ = grade(PolynomialRing(G.field, "X" => 1:n)[1])
+        z.reynolds_operator = reynolds_v_slm
+        I, M = proj_of_image_ideal(G, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(G, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, G, z.poly_ring, M, I, z.reynolds_operator)
+        return z
+    end
+    
+    function InvariantRing(R::RepresentationReductiveGroup, ring::MPolyRing)
+        n = ncols(R.rep_mat)
+        n == ngens(ring) || error("The given polynomial ring is not compatible.")
+        z = new()
+        z.representation = R
+        z.group = R.group
+        G = R.group
+        z.field = G.field
+        if is_graded(ring)
+            z.poly_ring = ring
+        else
+            z.poly_ring = grade(ring)
+        end
+        z.reynolds_operator = reynolds_v_slm
+        I, M = proj_of_image_ideal(G, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(G, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, G, z.poly_ring, M, I, z.reynolds_operator)
+        return z
+    end
+end
+
+invariant_ring(R::RepresentationReductiveGroup) = InvariantRing(R)
+fundamental_invariants(R::InvariantRing) = R.fundamental
+null_cone_ideal(R::InvariantRing) = R.NullConeIdeal
+
+function Base.show(io::IO, R::InvariantRing) #TODO compact printing
+    #if get(io, :supercompact, false)
+        print(io, "Invariant Ring of", "\n")
+        show(io, R.poly_ring)
+        print(io, " under group action of ", R.group.group[1], R.group.group[2], "\n", "\n")
+        print(io, "Generated by ", "\n")
+    join(io, R.fundamental, "\n")
+    #else
+     #   print(io, "Invariant Ring under ")
+      #  show(io, R.group[1], R.group[2])
+    #end
+end
+
+#I think this should work for everyone except tensor products.
+function image_ideal(G::ReductiveGroup, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
+    R = base_ring(rep_mat)
+    n = ncols(rep_mat)
+    m = G.group[2]
+    #r = G.direct_product[2]
+    mixed_ring_xy, x, y, zz = PolynomialRing(G.field, "x"=>1:n, "y"=>1:n, "zz"=>1:m^2)
+    # naming the variables zz here and z in the group ring
+    #for determinant - 
+    #M1 = matrix(mixed_ring_xy,m,m,[zz[i,j] for i in 1:m for j in 1:m])
+    ztozz = hom(R,mixed_ring_xy, gens(mixed_ring_xy)[(2*n)+1:(2*n)+(m^2)])
+    genss_ = gens(G.group_ideal)
+    genss = [ztozz(genss_[i]) for i in 1:length(genss_)]
+    #rep_mat in the new ring
+    new_rep_mat = matrix(mixed_ring_xy,n,n,[ztozz(rep_mat[i,j]) for i in 1:n, j in 1:n])
+    new_vars = new_rep_mat*[x[i] for i in 1:n] 
+    ideal_vect = [y[i] - new_vars[i] for i in 1:n] 
```

Comment:
Isn't this just this?
```suggestion
    ideal_vect = y - new_vars
```